### PR TITLE
Add details v2: fetching details from an external bookinfo service to demonstrate egress rules

### PR DESCRIPTION
**Pull request by vadimeisenbergibm**
The goal: demonstrate egress rules using bookinfo. Add details-v2 deployment that will fetch Comedy of Errors details from https://www.googleapis.com. This is the only free https JSON book service service I found, that returns details of "Comedy of Errors".

The code uses hard-coded ISBN 0486424618.

The demo: 
1. Undeploy details-v1
2. Deploy details-v2
3. Access the productpage. Message _"Sorry, product details are currently unavailable for this book."_ will be displayed. This is because without defining egress rules, all the from-the-mesh-outward traffic is blocked in kubernetes.

screen shot 2018-01-08 at 2 31 13 pm

4. Define egress rule:
```
cat <<EOF | istioctl create -f -
apiVersion: config.istio.io/v1alpha2
kind: EgressRule
metadata:
  name: googleapis
  namespace: default
spec:
  destination:
      service: "*.googleapis.com"
  ports:
      - port: 443
        protocol: https
  use_egress_proxy: false
EOF
```
5. Access the product page. Now the details are displayed.
screen shot 2018-01-17 at 4 15 06 am

6. delete the egress rule: `istioctl delete egressrule googleapis -n default`
7. Access the product page. The error message is displayed again, the from-the-mesh-outward traffic is blocked again.

### CODE REVIEW

File: samples/bookinfo/src/details/details.rb [hunk 1]
```diff
@@ -63,4 +74,29 @@ def get_book_details(id)
     }
 end
 
+def fetch_details_from_external_service(isbn, id)
+    uri = URI.parse('http://api.bookmooch.com')
```
Comment (rshriram) [inline samples/bookinfo/src/details/details.rb]: This is a bit unrealistic.. Since most people use HTTPS.

**Comment by rshriram** [comment]
screenshot?

File: samples/bookinfo/src/details/details.rb (hunk 1 above)
Comment (vadimeisenbergibm) [reply samples/bookinfo/src/details/details.rb]: This is just an example to demonstrate egress rules in a simplest way possible. I will write a blog about HTTP, and also about TCP/Mongo, and also about HTTPS. The first one would be about HTTP and would be the simplest one to understand and demonstrate. Note that in this case api.bookmooch.com provides HTTP API only, so this particular example is rather realistic.

**Comment by vadimeisenbergibm** [comment]
@rshriram I have added a screenshot to the PR's description.

**Comment by linsun** [comment]
hey @vadimeisenbergibm is it possible to use an IBM cloud service as an example such as the MySql service?  that way, you can show people best practices when using https with credentials.

**Comment by linsun** [comment]
FYI - https://console.bluemix.net/catalog/?search=mysql MySQL db service has a free tier.

**Comment by vadimeisenbergibm** [comment]
@linsun Good idea https://github.com/istio/istio/pull/2340#issuecomment-356019056, for demonstrating TCP egress rules.

**Comment by vadimeisenbergibm** [comment]
@rshriram @frankbu Following our discussion about http external service being unrealistic - I have implemented a bookinfo example of an https external book details service. Could you please review it ASAP? I would like to add it to Istio 0.5.0 and to write a blog post about it.

Note that in Ruby net/http package, the http/https client code is neatly structured, an option in the http client can be set to make a request http or https. The port 443 is derived automatically from the schema, so no need to specify it in the code, making the code natural.

```ruby
unless ENV['WITH_ISTIO'] === 'true' then
       http.use_ssl = true
end
```

File: samples/bookinfo/src/details/details.rb [hunk 2]
```diff
@@ -63,4 +75,46 @@ def get_book_details(id)
...
 end
 
+def fetch_details_from_external_service(isbn, id)
+    uri = URI.parse('https://www.googleapis.com/books/v1/volumes?q=isbn:' + isbn)
+    http = Net::HTTP.new(uri.host, uri.port)
+    http.read_timeout = 5 # seconds
+
+    unless ENV['WITH_ISTIO'] === 'true' then
+      http.use_ssl = true
+    end
```
Comment (frankbu) [inline samples/bookinfo/src/details/details.rb]: What does this do? It seems that you are passing `WITH_ISTIO: true` in the deployment, so the `unless` evaluates to `false`. So it's a NO-OP, right? The env variable name, `WITH_ISTIO`, doesn't have an obvious meaning, unless I'm just not understanding.

Comment (vadimeisenbergibm) [reply samples/bookinfo/src/details/details.rb]: `WITH_ISTIO` means that the code runs with istio sidecar proxy injected in the same pod. When it is injected, it is no-op. To run the code without Istio sidecar proxy injected, the user can undefine `WITH_ISTIO`.

Possible alternative names for `WITH_ISTIO`: `WITH_ISTIO_INJECTED`, `WITH_ISTIO_SIDECAR_INJECTED`, `ISTIOFIED`. I am open to any name proposals.

Comment (frankbu) [reply samples/bookinfo/src/details/details.rb]: Maybe just add a comment in the code to explain why `http.use_ssl` needs to be set when running without Istio, but it isn't needed when Istio sidecars are injected.

Comment (vadimeisenbergibm) [reply samples/bookinfo/src/details/details.rb]: @frankbu I have added the comment.

**Pull request merged**